File: crates/benches/fixtures/query_engine.go
package query

import (
	"fmt"
	"sort"
	"strings"
)

// Operator is a filter comparison operator.
type Operator string

const (
	OpEq         Operator = "eq"
	OpNeq        Operator = "neq"
	OpContains   Operator = "contains"
	OpStartsWith Operator = "starts_with"
	OpEndsWith   Operator = "ends_with"
)

// Filter describes a single filter predicate.
type Filter struct {
	Field    string
	Op       Operator
	Value    string
}

// SortOrder describes an ordering clause.
type SortOrder struct {
	Field     string
	Ascending bool
}

// QueryPlan is a structured query descriptor.
type QueryPlan struct {
	Filters []Filter
	Sort    []SortOrder
	Limit   int
	Offset  int
}

// QueryEngine executes queries over a slice of string maps.
type QueryEngine struct {
	records []map[string]string
}

// New creates a QueryEngine over the given records.
func New(records []map[string]string) *QueryEngine {
	return &QueryEngine{records: records}
}

// Execute runs the plan and returns matching records.
func (e *QueryEngine) Execute(plan QueryPlan) []map[string]string {
	result := make([]map[string]string, 0)
	for _, rec := range e.records {
		if matchesAll(rec, plan.Filters) {
			result = append(result, rec)
		}
	}
	if len(plan.Sort) > 0 {
		sort.SliceStable(result, func(i, j int) bool {
			for _, s := range plan.Sort {
				vi, vj := result[i][s.Field], result[j][s.Field]
				if vi == vj { continue }
				if s.Ascending { return vi < vj }
				return vi > vj
			}
			return false
		})
	}
	if plan.Offset > 0 && plan.Offset < len(result) { result = result[plan.Offset:] }
	if plan.Limit > 0 && plan.Limit < len(result) { result = result[:plan.Limit] }
	return result
}

func matchesAll(rec map[string]string, filters []Filter) bool {
	for _, f := range filters {
		v, ok := rec[f.Field]
		if !ok { return false }
		switch f.Op {
		case OpEq:         if v != f.Value { return false }
		case OpNeq:        if v == f.Value { return false }
		case OpContains:   if !strings.Contains(v, f.Value) { return false }
		case OpStartsWith: if !strings.HasPrefix(v, f.Value) { return false }
		case OpEndsWith:   if !strings.HasSuffix(v, f.Value) { return false }
		default:           return false
		}
	}
	return true
}

// Count returns the number of records matching the filters.
func (e *QueryEngine) Count(filters []Filter) int {
	n := 0
	for _, rec := range e.records {
		if matchesAll(rec, filters) { n++ }
	}
	return n
}

// Distinct returns unique values for a field.
func (e *QueryEngine) Distinct(field string) []string {
	seen := make(map[string]bool)
	var result []string
	for _, rec := range e.records {
		if v, ok := rec[field]; ok && !seen[v] {
			seen[v] = true
			result = append(result, v)
		}
	}
	sort.Strings(result)
	return result
}

// Explain returns a human-readable description of the plan.
func (e *QueryEngine) Explain(plan QueryPlan) string {
	return fmt.Sprintf("filters=%d sort=%d limit=%d offset=%d records=%d", len(plan.Filters), len(plan.Sort), plan.Limit, plan.Offset, len(e.records))
}

File: crates/benches/fixtures/query_engine.py
from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum, auto
from typing import Any, Callable, Optional


class Operator(Enum):
    EQ = auto()
    NEQ = auto()
    LT = auto()
    LTE = auto()
    GT = auto()
    GTE = auto()
    CONTAINS = auto()
    STARTS_WITH = auto()
    ENDS_WITH = auto()


@dataclass
class Filter:
    field: str
    operator: Operator
    value: Any

    def matches(self, record: dict[str, Any]) -> bool:
        field_val = record.get(self.field)
        if field_val is None:
            return False
        match self.operator:
            case Operator.EQ: return field_val == self.value
            case Operator.NEQ: return field_val != self.value
            case Operator.LT: return field_val < self.value
            case Operator.LTE: return field_val <= self.value
            case Operator.GT: return field_val > self.value
            case Operator.GTE: return field_val >= self.value
            case Operator.CONTAINS: return self.value in str(field_val)
            case Operator.STARTS_WITH: return str(field_val).startswith(self.value)
            case Operator.ENDS_WITH: return str(field_val).endswith(self.value)
        return False


@dataclass
class QueryPlan:
    filters: list[Filter]
    order_by: Optional[str] = None
    descending: bool = False
    limit: Optional[int] = None
    offset: int = 0


class QueryEngine:
    def __init__(self, records: list[dict[str, Any]]) -> None:
        self._records = records

    def execute(self, plan: QueryPlan) -> list[dict[str, Any]]:
        result = [r for r in self._records if all(f.matches(r) for f in plan.filters)]
        if plan.order_by:
            result.sort(key=lambda r: r.get(plan.order_by, None), reverse=plan.descending)
        result = result[plan.offset:]
        if plan.limit is not None:
            result = result[:plan.limit]
        return result

    def count(self, filters: list[Filter]) -> int:
        return sum(1 for r in self._records if all(f.matches(r) for f in filters))

    def distinct(self, field: str) -> list[Any]:
        seen: set[Any] = set()
        result = []
        for r in self._records:
            v = r.get(field)
            if v not in seen:
                seen.add(v)
                result.append(v)
        return result
